### PR TITLE
ipldutil: simplify state synchronization, add docs

### LINKED ISSUE #274: Document ipldutil.Traverser
# What

Add comments on private functions and unclear blocks of code in https://github.com/ipfs/go-graphsync/blob/main/ipldutil/traverser.go to help explain intent

There are already comments on most public functions, but the private functions don't have docs and there are large blocks of code that may be confusing.

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

Fixes #274.

### CODE REVIEW

**Comment by mvdan** [comment]
I've stared at the code for a while, and I think understanding the TODO bits will help unblock me to document the rest of the code.

File: ipldutil/traverser.go
Comment (hannahhoward) [inline ipldutil/traverser.go:138]: Concur on the complexity cause I had to reread my code to rediscover the intent. But here it is:

From the standpoint of the consumer of the api, IsComplete and CurrentRequest give you information about the state of the traversal.

The state is cached 4 values:
- isDone = is everything wrapped up
- completionError - if it finished, did everything go ok?
- currentLink - what am I supposed to load next
- currentContext - waht is the link context for the next load

When I call `Advance` or `Error`, it sends a response to the loader which allows the traversal to continue. All of these values that are cached are no longer valid for the next call to IsComplete or CurrentRequest. Putting a value in awaitRequest essentially indicates on the next call to IsComplete or CurrentRequest, you need to wait and update the state. This is done by reading from the stateChan.

As I look at this, I agree that there are concurrency hickups here -- If the consumer interface were used by multiple threads for the same traversal there could be race conditions if two different threads called IsComplete/CurrentRequest around the same time when awaitRequest had a value. As it is, this hasn't come up yet cause there's only ever a single thread consuming, but we should fix it. I wonder if we should turn to my new favorite friend sync.Cond somehow (I've become way to enamored of it's Broadcast capabilites). But if there are good ways to simplify this period, I agree it's very confusing.

File: ipldutil/traverser.go
Comment (hannahhoward) [reply ipldutil/traverser.go:250]: https://github.com/ipfs/go-graphsync/blob/main/responsemanager/queryexecutor/queryexecutor.go#L191-L193

**Review by hannahhoward — COMMENTED** [review]
Thanks this is helpful! Will reach out about next steps

File: ipldutil/traverser.go
Comment (rvagg) [inline ipldutil/traverser.go:26]: It'd be nice to see if that can be completely resolved by doing some global replacing. Related to this is https://github.com/ipfs/go-graphsync/pull/258 where you can see an original TODO in the diff that was partially resolved.

**Comment by rvagg** [comment]
Is the intent to resolve the TODOs on this branch, or do them piecemeal in future updates? Other than that comment block that Hannah has provided some clarity on this lgtm since it doesn't look like it has material impact.

File: ipldutil/traverser.go
Comment (mvdan) [reply ipldutil/traverser.go:99]: Just to clarify that the values are only used as "else" fallbacks. I initially thought that we always used dagpb.AddSupportToChooser, for instance.

File: ipldutil/traverser.go
Comment (mvdan) [reply ipldutil/traverser.go:250]: correct, I was just looking at one file - oops.

File: ipldutil/traverser.go
Comment (mvdan) [reply ipldutil/traverser.go:26]: @masih @hannahhoward thoughts? happy to replace all uses with `errors.Is` and `context.Cancelled` directly, if it sounds good.

Comment (mvdan) [reply ipldutil/traverser.go:26]: Mental note: the only reason we might not want to use `context.Cancelled` is if our `ContextCancelError` type wasn't equivalent, e.g. if it doesn't always replace `context.Cancelled`. Though, arguably, then it should have a different name to clarify when it's used and when it's not versus the standard error.

File: ipldutil/traverser.go
Comment (mvdan) [reply ipldutil/traverser.go:138]: Thanks for the extra context! It's all clear now.

File: ipldutil/traverser.go
Comment (hannahhoward) [reply ipldutil/traverser.go:26]: I think we could? I authorize this refactor! 😛

**Comment by mvdan** [comment]
@rvagg apologies for the late reply - the TODOs were open questions for Hannah, really. The intent was to have them during the early stages of code review so I could understand how the code worked.

**Comment by mvdan** [comment]
@hannahhoward this should be ready for review and/or testing. Please pay extra attention to how the mutex is documented and used, in particular to the `loader` method.

**Comment by mvdan** [comment]
Friendly bump that this is ready for review/merge :)

File: ipldutil/traverser.go
Comment (hannahhoward) [inline ipldutil/traverser.go:269]: should we unlock in this return block? What happens if Advance gets called twice after the traversal completes? Is that not a deadlock then?

File: ipldutil/traverser.go
Comment (hannahhoward) [inline ipldutil/traverser.go:274]: Same issue here -- if the context is done, what will cause the lock to ever unlock again?

File: ipldutil/traverser.go
Comment (hannahhoward) [inline ipldutil/traverser.go:286]: see commentary above about unlocking for done

**Review by hannahhoward — CHANGES_REQUESTED** [review]
I'm concerned about deadlocking on multiple calls to Advance/Error after the traversal is done. I mean generally this shouldn't happen anyway, but I think you want to release the lock unless you are actually sending a response over the nextResponse channel

File: ipldutil/traverser.go
Comment (mvdan) [reply ipldutil/traverser.go:271]: Done, with a test covering all four methods called post-done.

**Comment by mvdan** [comment]
Should be ready for a final review+merge I think :)

**Comment by mvdan** [comment]
Seems like a flake on TestNetworkDisconnect. Does that ring a bell to either of you? I don't imagine my change is the cause.

**Review by hannahhoward — APPROVED** [review]

**Pull request merged**
